crates/stc_ts_file_analyzer/src/analyzer/expr/mod.rs

```diff
@@ -1048,7 +1048,38 @@ impl Analyzer<'_, '_> {
 
         matching_elements.dedup_type();
 
-        Ok(Some(Type::union(matching_elements)))
+        let mut res_vec = vec![];
+
+        for el in matching_elements.into_iter() {
+            if let Ok(res) = self.normalize(Some(span), Cow::Owned(el), Default::default()) {
+                res_vec.push(res.into_owned());
+            }
+        }
+        res_vec.dedup_type();
+        if res_vec.len() == 1 {
+            return Ok(res_vec.pop());
+        }
+        let result = match type_mode {
+            TypeOfMode::LValue => {
+                let mut temp_vec = vec![];
```

Comment:
I think `Type::new_intersection(span, res_vec)` will do what you are trying to do

Reply:
And more efficient

Reply:
Seems like this is not fixed

Reply:
I meant 1064 ~ 1072 can be replaced as a whole.
`new_intersection` has logic for 

```rust
                for elem in res_vec {
                    if let Type::Intersection(ty) = elem.normalize() {
                        temp_vec.extend(ty.types.iter().cloned());
                    } else {
                        temp_vec.push(elem);
                    }
                }
                temp_vec.dedup_type();
```

Reply:
Oh, the next code was in new_intersection
```rs
for ty in iter {
    if ty.is_intersection() {
        tys.extend(ty.expect_intersection().types);
    } else {
        tys.push(ty);
    }
}
tys.dedup_type();
```

I will replace it

thx!

Reply:
I fixed it